interpreter/cling/include/cling/Interpreter/Value.h

```diff
@@ -129,21 +188,18 @@ namespace cling {
         EStorageType storageType = V.getStorageType();
         switch (storageType) {
         case kSignedIntegerOrEnumerationType:
-          return (T) V.getAs<long long>();
         case kUnsignedIntegerOrEnumerationType:
-          return (T) V.getAs<unsigned long long>();
         case kDoubleType:
-          return (T) V.getAs<double>();
         case kFloatType:
-          return (T) V.getAs<float>();
         case kLongDoubleType:
-          return (T) V.getAs<long double>();
+          return (T) V.getAs<T>();
```

Comment:
If the value was stored as `int` but here we are asked to read it as `char` then this will give broken values on big endian because the cast from `int` to `char` isn't happening.

Reply:
In that case `getAs<T>` will look like that:
```
char getAs<char> () {
  switch (BT->getKind()) {
    ...
    case BultinType::Int: (char) m_Storage.m_Int;
    ...
  }
}
```
That allows us to access the type representation the way it was set and also let the compiler see to what type we want to cast it to generate such code. We have a test called `Conversions`. I can add that test there, too...

Reply:
Fair enough.  I agree it is helpful to move the functionality and I think it can be improved to also keep the previous performance; this is important as the previous `getAs` is used in performance critical code.

If I understood correctly the previous `getAs` was literally just returning the address:
```
    void*& getAs(double*) { return m_Storage.m_Ptr; } // reach via a template overload function
```
while the new getAs does at least (in addition to the lookups of the search for address)
* one (llvm) dynamic_cast
* 3 lookups (that's assuming all the function calls are inlined)
to get to the kind and then
* one (large-ish) switch statement to find the value
* one conversion (unless you ask for the right type.

There are some performance critical routines (eg `setValueNoAlloc` which already knows the type of the contained value and for them the extra cost in the proposed `getAs` is spurrious.   I guess that most of the performance could be retrieve by caching/recording the kind inside the value but it still "slower" than just returning the address.

I.e. I can see the need for 2 distincts version, one that return the (address of) the value "as-requested" by the user and one that does all the necessary conversions.

In the existing code we have:
* `getAs` : returns the underlying address/value as requested by the user, no checks, no conversion
* `simplisticCastAs`: returns the value after undergoing the requested _numerical_ conversions.

While in the PR code we have both `getAs` and `simplisticCastAs` that return the value after under going the request _numerical_ conversions.  I.e. there is no fast path left (more exactly no 'templated' fast path, there is still (for now?) access to a (now not complete set of) routines spelt explicitly: `getDouble`, `getLongDouble`, etc.

So to retrieve the fast path, we would probably need to extend the `get[TYPENAME]` to cover all the cases.

But also for programming convenience we should re-introduce a templated wrapper around those.  i.e. having 2 templates:
(a) returns the underlying address/value as requested by the user, no checks, no conversion
(b) returns the value after undergoing the requested _numerical_ conversions.

and we would need to find names for it.   One easy proposal is to return to the name used up to now (`getAs` would have to return to being a simple wrapper around `get[TYPENAME]` while `simplisticCastAs` would be implemented in term of `convert`).

Alternatively we could find a new name for `returns the underlying address/value as requested by the user, no checks, no conversion`

Cheers,
Philippe.

PS. Another "interface change" is that the previous `getAS` returns a reference to the underlying value/storage and thus can be used for assignment while the new `getAs` returns by value and thus change semantic dramatically (hopefully the compiler will about assignment to a temporary value) ... and https://github.com/root-project/root/pull/11252/files#r965189935 is worrisome.

Reply:
> Fair enough. I agree it is helpful to move the functionality and I think it can be improved to also keep the previous performance; this is important as the previous `getAs` is used in performance critical code.
> 
> If I understood correctly the previous `getAs` was literally just returning the address:
> 
> ```
>     void*& getAs(double*) { return m_Storage.m_Ptr; } // reach via a template overload function
> ```

  I believe these should be improved. The `getAs` should not be returning an address, we should have a dedicated set of overloads for this. However, I'd keep that outside of the current PR.

> 
> while the new getAs does at least (in addition to the lookups of the search for address)
> 
>     * one (llvm) dynamic_cast
> 
>     * 3 lookups (that's assuming all the function calls are inlined)
>       to get to the kind and then
> 
>     * one (large-ish) switch statement to find the value
> 
>     * one conversion (unless you ask for the right type.
> 
> 
> There are some performance critical routines (eg `setValueNoAlloc` which already knows the type of the contained value and for them the extra cost in the proposed `getAs` is spurrious. I guess that most of the performance could be retrieve by caching/recording the kind inside the value but it still "slower" than just returning the address.


My expectation is that will have close to 0 overhead as the `dyn_cast` compares two `int`s and the switch stmt will be compiled into a simple jump table. I will do some performance benchmarks on this. 

> 
> I.e. I can see the need for 2 distincts version, one that return the (address of) the value "as-requested" by the user and one that does all the necessary conversions.
> 
> In the existing code we have:
> 
>     * `getAs` : returns the underlying address/value as requested by the user, no checks, no conversion
> 
>     * `simplisticCastAs`: returns the value after undergoing the requested _numerical_ conversions.
> 
> 
> While in the PR code we have both `getAs` and `simplisticCastAs` that return the value after under going the request _numerical_ conversions. I.e. there is no fast path left (more exactly no 'templated' fast path, there is still (for now?) access to a (now not complete set of) routines spelt explicitly: `getDouble`, `getLongDouble`, etc.


I'd prefer not to do that if the performance impact is negligible. This is a source of buggy code. Maybe we could make some of these well known cases friends and keep these interfaces private?

> 
> So to retrieve the fast path, we would probably need to extend the `get[TYPENAME]` to cover all the cases.
> 
> But also for programming convenience we should re-introduce a templated wrapper around those. i.e. having 2 templates: (a) returns the underlying address/value as requested by the user, no checks, no conversion (b) returns the value after undergoing the requested _numerical_ conversions.
> 
> and we would need to find names for it. One easy proposal is to return to the name used up to now (`getAs` would have to return to being a simple wrapper around `get[TYPENAME]` while `simplisticCastAs` would be implemented in term of `convert`).

Yes, perhaps we can make `get[TYPENAME]` private and then `simplisticCastAs` becomes `castAs` (as this seems to be a common notion in llvm).

> 
> Alternatively we could find a new name for `returns the underlying address/value as requested by the user, no checks, no conversion`

I'd prefer to use a setter rather than returning an address.

> 
> Cheers, Philippe.
> 
> PS. Another "interface change" is that the previous `getAS` returns a reference to the underlying value/storage and thus can be used for assignment while the new `getAs` returns by value and thus change semantic dramatically (hopefully the compiler will about assignment to a temporary value) ... and https://github.com/root-project/root/pull/11252/files#r965189935 is worrisome.

Yep, another case for going for a setter.

Reply:
Here is a standalone example of the approach we have right now: https://godbolt.org/z/M7bYKG1zG

The assembly comes out fairly optimized and there is no difference if get[TYPENAME] was called or we did extra operations to cast. Maybe you can provide an example where this is not the case?

Reply:
It is a little 'too' standalone :).  As the compiler knows both the input and the output it can (and does) optimize out all the unused path.  Maybe https://godbolt.org/z/3ea4YWxPs is a little bit more realistic.

Also in this PR's `getAs`, the more concerning part is:
```
    if (const auto *ET = llvm::dyn_cast<clang::EnumType>(QT.getTypePtr()))
      QT = ET->getDecl()->getIntegerType();

    QT = QT.getCanonicalType();

    assert(QT->isBuiltinType());

    const auto* BT = llvm::cast<const clang::BuiltinType>(QT.getTypePtr());
/// then get BT->getKind()
```
and there is no way, it is as efficient as : 
```
return datamembername;
```
(the switch statement itself, indeed seems to be just a jump table .. but it is still extra :) ).  We will see with your measurement how much that might matter.

Reply:
It looks pretty good.  Thanks.  (For the record, small issue left is understanding better if `getPtrAddress` is the correct implementation and an odd fixme about assert and `setLongLong`)